packages/paymaster/src/BiconomyPaymaster.ts

```diff
@@ -1,49 +1,246 @@
+import { ERC20_ABI } from './constants'
+import { IHybridPaymaster } from './interfaces/IHybridPaymaster'
 
-export class BiconomyPaymaster implements IPaymaster {
+// Hybrid - Generic Gas Abstraction paymaster
+/**
+ *
+ */
+export class BiconomyPaymaster implements IHybridPaymaster {
   constructor(readonly paymasterConfig: PaymasterConfig) {}
+
+  // Note: maybe rename to createTokenApprovalTransaction
+  async createTokenApprovalRequest(
+    tokenPaymasterRequest: BiconomyTokenPaymasterRequest,
+    provider: Provider
+  ): Promise<Transaction> {
+    const feeTokenAddress: string = tokenPaymasterRequest.feeQuote.tokenAddress
+    Logger.log('erc20 fee token address ', feeTokenAddress)
+
+    const spender = tokenPaymasterRequest.spender
+    Logger.log('spender address ', spender)
+
+    // TODO
+    // Note: should also check in caller if the approval is already given, if yes return object with address or data 0
+    // Note: we would need userOp here to get the account/owner info to check allowance
+
+    let requiredApproval: BigNumberish = BigNumber.from(0)
+
+    if (tokenPaymasterRequest.maxApproval && tokenPaymasterRequest.maxApproval == true) {
+      requiredApproval = ethers.constants.MaxUint256
+    } else {
+      /*requiredApproval = ethers.BigNumber.from(
+        tokenPaymasterRequest.feeQuote.maxGasFee.toString()
+      ).mul(
+        ethers.BigNumber.from(10).pow(
+          ethers.BigNumber.from(tokenPaymasterRequest.feeQuote.decimal.toString())
+        )
+      )*/
+
+      requiredApproval = BigNumber.from(
+        Math.floor(
+          tokenPaymasterRequest.feeQuote.maxGasFee *
+            Math.pow(10, tokenPaymasterRequest.feeQuote.decimal)
+        )
+      )
+    }
+
+    Logger.log('required approval for erc20 token ', requiredApproval)
+
+    const erc20 = new ethers.Contract(feeTokenAddress, ERC20_ABI, provider)
+
+    // TODO?
+    // Note: For some tokens we may need to set allowance to 0 first so that would return batch of transactions and changes the return type to Transaction[]
+    // In that case we would return two objects in an array, first of them being..
+    /*
+    {
+      to: erc20.address,
+      value: ethers.BigNumber.from(0),
+      data: erc20.interface.encodeFunctionData('approve', [spender, BigNumber.from("0")])
+    }
+    */
+
+    return {
+      to: erc20.address,
+      value: ethers.BigNumber.from(0),
+      data: erc20.interface.encodeFunctionData('approve', [spender, requiredApproval])
+    }
+  }
+
   /**
    *
    * @param userOp
-   * @description This function will send userOp to paymaster service to get signed from paymaster signer
-   * @returns string
+   * @param paymasterServiceData
+   * @returns
    */
-  async getPaymasterAndData(userOp: Partial<UserOperation>): Promise<string> {
+  async getPaymasterFeeQuotesOrData(
+    userOp: Partial<UserOperation>,
+    paymasterServiceData: FeeQuotesOrDataDto
+  ): Promise<BiconomyTokenPaymasterFeeQuoteResponse | string> {
```

Comment:
Should avoid returning two different types of response from a method. Create a single response type and mode field in the response and other optional parameters.

So developer don't need to do type checking on the result before using them. Developer will know to check mode and then decide which params to expect in response.